### PR TITLE
Provide ability to delete contacts

### PR DESCRIPTION
**Pull request by base10**
- Refactoring
- Show a message when contact list is empty

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -28,6 +28,7 @@ $(document).ready(function() {
 
   $("#new_contact").click( function() {
     hideFeedback();
+    showContactForm();
```
Comment (base10) [inline app/assets/javascripts/application.js:31]: It's there, just not in this file.

File: app/assets/javascripts/contacts.js [hunk 2]
```diff
@@ -43,27 +45,38 @@ var showSuccess = function(contact, method) {
   );
 };
 
-var erredContacts = function(_xhr, status, error) {
+var recordError = function(xhr, status, error, message) {
```
Comment (drapergeek) [inline app/assets/javascripts/contacts.js]: Minor: `recordError` seems very "passive" to me. If I saw this function name I would assume it was just tracking it and not that anything was presented to the user.

File: app/assets/javascripts/contacts.js [hunk 3]
```diff
@@ -72,8 +85,101 @@ var processContacts = function(contacts, status, xhr) {
...
 };
 
-var showContactInForm = function(position) {
+var showEmptyContactList = function() {
+  $(".contacts").append(
+    "<li data-role='contact'>No contacts yet</li>"
+  );
+};
+
+var processContacts = function(contacts, status, xhr) {
```
Comment (drapergeek) [inline app/assets/javascripts/contacts.js]: We use the convention of using leading underscores for variables you don't use (`(contacts, _status, _xhr)`). But in JS I typically just leave them out unless I need one later in the chain (I need xhr and I don't need status for example).

File: app/assets/javascripts/contacts.js [hunk 4]
```diff
@@ -72,8 +85,101 @@ var processContacts = function(contacts, status, xhr) {
...
+      retrieveContacts,
+      showSuccess(contact, method)
+    ],
+    url: determineContactUrl(contact.id),
+  });
+};
+
+var deleteContact = function(position) {
+  var contact = storedContacts[position];
+  var confirmation = confirm("Verify you want to delete " + contact.name);
```
Comment (base10) [inline app/assets/javascripts/contacts.js:157]: confirm() is a global.

File: app/assets/javascripts/contacts.js [hunk 5]
```diff
@@ -72,8 +85,101 @@ var processContacts = function(contacts, status, xhr) {
...
+  $(".contacts").append(
+    "<li data-role='contact'>No contacts yet</li>"
+  );
+};
+
+var processContacts = function(contacts, status, xhr) {
+  storedContacts = contacts;
+  $("[data-role='contact']").remove();
+
+  if (anyContacts()) {
```
Comment (drapergeek) [inline app/assets/javascripts/contacts.js:95]: I think I would move the concept of showing an empty or full contact list to its own function.

File: app/assets/javascripts/contacts.js (hunk 4 above)
Comment (drapergeek) [inline app/assets/javascripts/contacts.js:157]: You can add it to the jshint to let hound know.

Comment (base10) [inline app/assets/javascripts/contacts.js:157]: Addressed with a3434af

File: app/assets/javascripts/contacts.js [hunk 6]
```diff
@@ -85,9 +185,22 @@ var showContactInForm = function(position) {
   $("[data-role='notes']").val(contact.notes);
 };
 
+var resetButtons = function(position) {
```
Comment (base10) [inline app/assets/javascripts/contacts.js:183]: Addressed with c08e6417

File: app/assets/javascripts/contacts.js (hunk 5 above)
Comment (base10) [inline app/assets/javascripts/contacts.js:95]: Addressed with 26bb677

File: app/assets/javascripts/contacts.js [hunk 7]
```diff
@@ -43,27 +45,38 @@ var showSuccess = function(contact, method) {
...
-  showError("Adding or updating the contact");
+  recordAndDisplayError(xhr, status, error, "Adding or updating the contact");
 };
 
-var processContacts = function(contacts, status, xhr) {
-  storedContacts = contacts;
-  $("[data-role='contact']").remove();
 
+var anyContacts = function() {
+  if (storedContacts.length > 0) {
```
Comment (drapergeek) [inline app/assets/javascripts/contacts.js:71]: Can't this just return `storedContacts.length > 0`?

File: app/assets/javascripts/contacts.js [hunk 8]
```diff
@@ -72,8 +85,86 @@ var processContacts = function(contacts, status, xhr) {
...
+  if (anyContacts()) {
+    showContactList();
+  }
+  else {
+    showEmptyContactList();
+  }
+};
+
+var processContacts = function(contacts) {
+  storedContacts = contacts;
```
Comment (drapergeek) [inline app/assets/javascripts/contacts.js:104]: What is this? It looks like you have assigned something but it isn't declared as a variable and you don't seem to use it anywhere.

File: app/assets/javascripts/contacts.js (hunk 7 above)
Comment (base10) [inline app/assets/javascripts/contacts.js:71]: Trying it out, no, unfortunately.

File: app/assets/javascripts/contacts.js (hunk 8 above)
Comment (base10) [inline app/assets/javascripts/contacts.js:104]: Declared on line 5, used in anyContacts(), showContactList(), elsewhere

Comment (drapergeek) [inline app/assets/javascripts/contacts.js:104]: Ah. This seems confusing. I might move this to a function so you have something like "setContactList(contacts)". Not sure though, do whatever feels right.

Comment (base10) [inline app/assets/javascripts/contacts.js:104]: Addressed with 9aaf8fe

**Comment by drapergeek** [comment]
:shipit:

**Pull request merged**